### PR TITLE
Submiting my version of app

### PR DESCRIPTION
**Pull request by Mindaugas-Jacionis**
I've added my comments on Readme.md to provide all needed on info on my solution. :)

### CODE REVIEW

File: client/src/reducers/servers/reducer.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    case types.SUCCESS:
+      return Object.assign({}, state, {
+        isFetching: false,
+        errorMessage: false,
+        servers: payload
+      });
+    case types.FAILURE:
+      return Object.assign({}, state, {
+        isFetching: false,
+        errorMessage: 'Something went horibly wrong, please try later!'
```
Comment (tmxyz) [inline client/src/reducers/servers/reducer.js]: Grammar: horribly

File: client/src/reducers/login/reducer.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+      return Object.assign({}, state, {
+        isFetching: false,
+        errorMessage: message || false,
+        token: token
+      });
+    case types.TOKEN_FAILURE:
+      return Object.assign({}, state, {
+        isFetching: false,
+        errorMessage: 'Something went wrong while login in'
```
Comment (tmxyz) [inline client/src/reducers/login/reducer.js]: Grammar: logging

File: client/src/__tests__/reducers.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import configureStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import { ApiMiddleware } from '../utils';
+import * as loginActions from '../reducers/login/actions';
+import * as serversActions from '../reducers/servers/actions';
+import loginReducer from '../reducers/login/reducer';
+import * as loginTypes from '../reducers/login/actionTypes';
+
+const middlewares = [thunk, ApiMiddleware('http://playground.tesonet.lt/v1')];
```
Comment (tmxyz) [inline client/src/__tests__/reducers.js]: Move `http://playground.tesonet.lt/v1` to a separate file and make it a constant.

File: client/src/components/ui/Button.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    onClick();
+  }
+
+  render() {
+    const { type, text } = this.props;
+
+    return (
+      <button
+        type={type}
+        onClick={() => this.onClick()}
```
Comment (Mindaugas-Jacionis) [inline client/src/components/ui/Button.js]: Consistency - always having onClick/onChang/onSubmit/.etc as a separate method helps to navigate files and code in them when they follow same structure. :)

File: client/src/components/ui/Input.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      value: '',
+      focus: false
+    }
+  }
+
+  onChange(event) {
```
Comment (Mindaugas-Jacionis) [inline client/src/components/ui/Input.js]: Habit from React Native. In RN if you do not retrieve val from props you fall back to state, but in order to have value in state you need to set it there on `onChange`. In this case - this helps handling icons(especially with browser autofill).

File: client/src/__tests__/components.js [hunk 6]
```diff
@@ -39,9 +40,53 @@ test('Basic ServerRow renders without crash', () => {
...
+  // manually trigger the onClick
+  tree.props.onSubmit();
+  // re-rendering
+  tree = component.toJSON();
+  expect(tree).toMatchSnapshot();
+});
+
+/* Testing props of components */
+describe('Component prop tests', () => {
+  it('should render Button with text & type props', done => {
```
Comment (tmxyz) [inline client/src/__tests__/components.js]: You are not doing anything aync in this test, so no need to have done here.

File: client/src/components/login/LoginForm.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    super(props);
+    this.state = {
+      username: '',
+      password: ''
+    }
+  }
+
+  validateForm() {
+    const data = this.state;
+    const dataObjValues = Object.keys(data).map(
```
Comment (tmxyz) [inline client/src/components/login/LoginForm.js]: Missed this.

File: client/src/components/ui/Input.js [hunk 8]
```diff
@@ -47,20 +47,23 @@ class Input extends Component {
   }
 
   render() {
-    const { type, placeholder, value, icon } = this.props;
+    const { type, placeholder, value, icon, invalid } = this.props;
     const state = this.state;
     const { focus } = state;
     const showIcon = icon && icon !== '' && state.value === '' && !focus;
+    let className = ['input'];
```
Comment (tmxyz) [inline client/src/components/ui/Input.js]: Although I like that you are using third party libraries very sparingly, sometimes it is a good idea to use them. In this case `classnames` library would be perfect. If you don't want to use it, which is totally fine, get rid of the ugly `push`'es

File: client/src/index.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
```
Comment (tmxyz) [inline client/src/index.js]: I meant this:
```
import { render } from 'react-dom';

...

render(<App />, document.getElementById('root'));
```

File: client/src/screens/servers/ServersScreen.js [hunk 10]
```diff
@@ -91,13 +93,23 @@ class ServersScreen extends Component {
 function mapStateToProps(state, props) {
   const { servers, isFetching, errorMessage } = state.servers;
   const { token } = state.login;
+  let sortedServers = Immutable.asMutable(servers);
+  sortedServers = sortedServers.sort(
```
Comment (tmxyz) [inline client/src/screens/servers/ServersScreen.js]: `.sort()` modifies array in place, so no need to reasign it to itself. Also, why are you using `Immutable` here? Also, why are you using square bracket notation i.e.`a['distance']` to access the object properties?

File: client/src/utils/Constants.js [hunk 11]
```diff
@@ -0,0 +1 @@
+export const baseUrl = 'http://playground.tesonet.lt/v1';
```
Comment (tmxyz) [inline client/src/utils/Constants.js]: I'd prefer if constants be all uppercase e.g. `BASE_URL`.

File: server.js [hunk 12]
```diff
@@ -25,10 +25,10 @@ app.post('/tokens', (req, res) => {
     });
 });
 
-app.get("/servers", (req, res) => {
+app.get('/servers', (req, res) => {
   const getData = {
     headers: {
-      "authorization": req.headers.authorization
+      'authorization': req.headers.authorization
```
Comment (tmxyz) [inline server.js]: You don't need to surround `authorization` with quotes.

**Pull request closed without merging**